### PR TITLE
Fixed issue with leased connection leaks when threads executing HTTP …

### PR DESCRIPTION
**Pull request by joviegas**
Fixed issue with leased connection leaks when threads executing HTTP connections with Apache HttpClient were interrupted while the connection was in progress.

## Motivation and Context

- The Apache connection pool leases were not released when Thread  was interrupted while Connection was in progress.
- The fix will release the resources in case if the thread was interrupted during the get.

 ## Root cause
- When a Thread is interrupted externally while httpClient.execute() is in progress , it is not necessary that the Httpclient interrupts the threads  because of  http.Client execute call completes successfully .
- At a later stage in SDK Execute Pipeline the thread interrupt gets checked and throws execoption whithout closing or aborting the stream , this causes the Resource leaks.

## Modifications
- Close the response output stream in AfterTransmissionExecutionInterceptorsStage by passing response in checkInterrupted API.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Checklist

- [ ] I have read the [CONTRIBUTING](https://github.com/aws/aws-sdk-java-v2/blob/master/CONTRIBUTING.md) document
- [ ] Local run of `mvn install` succeeds
- [ ] My code follows the code style of this project
- [ ] My change requires a change to the Javadoc documentation
- [ ] I have updated the Javadoc documentation accordingly
- [ ] I have added tests to cover my changes
- [ ] All new and existing tests passed
- [ ] I have added a changelog entry. Adding a new entry must be accomplished by running the `scripts/new-change` script and following the instructions. Commit the new file created by the script in `.changes/next-release` with your changes.
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License

- [ ] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

**Comment by joviegas** [comment]
Had a internal discussion and we will be updating the Client core to close the stream rather than closing at SdkApache layer.
This is done to keep the contract of Apache client same as Sdk since we donot want to just patch it at SdkApache layer.

File: test/protocol-tests/src/test/java/software/amazon/awssdk/protocol/tests/connection/SyncClientConnectionInterruptionTest.java [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+        toBeInterruptedFuture0.cancel(true);
+        toBeInterruptedFuture1.cancel(true);
+        toBeInterruptedFuture2.cancel(true);
+        unInterruptedSleep(LONG_DELAY / 2);
+        // Make sure thread start the Http connections
+        AllTypesResponse allTypesResponse = client.allTypes();
+        assertThat(allTypesResponse.stringMember()).isEqualTo("resultString");
+        executorService.shutdownNow();
+    }
+
```
Comment (joviegas) [inline test/protocol-tests/src/test/java/software/amazon/awssdk/protocol/tests/connection/SyncClientConnectionInterruptionTest.java:110]: Will format the test to remove extra lines

File: .changes/next-release/bugfix-ApacheHTTPClient-bf0095f.json [hunk 2]
```diff
@@ -0,0 +1,6 @@
+{
+    "type": "bugfix",
+    "category": "Apache HTTP Client",
```
Comment (zoewangg) [inline .changes/next-release/bugfix-ApacheHTTPClient-bf0095f.json:3]: nit: this is not just applicable to Apache HTTP client right? Should we use `AWS SDK for Java v2` or whatever the right category for core?

File: test/protocol-tests/src/test/java/software/amazon/awssdk/protocol/tests/connection/SyncClientConnectionInterruptionTest.java [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+    @Test
+    void interruptionWhenWaitingForLease_AbortsImmediately() throws InterruptedException {
```
Comment (zoewangg) [inline test/protocol-tests/src/test/java/software/amazon/awssdk/protocol/tests/connection/SyncClientConnectionInterruptionTest.java:116]: How do we know the request is waiting for a connection from the pool when it gets aborted?

File: test/protocol-tests/src/test/java/software/amazon/awssdk/protocol/tests/connection/SyncClientConnectionInterruptionTest.java [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+        unInterruptedSleep(SERVER_RESPONSE_DELAY - SERVER_RESPONSE_DELAY / 10);
+        leaseWaitingThread.interrupt();
+        leaseWaitingThread.join();
+        assertThat(exception.getException()).isInstanceOf(ApiCallAttemptTimeoutException.class);
+        client.close();
+    }
+
+
+    @Test
+    void sdkClientInterrupted_while_connectionIsInProgress() throws InterruptedException {
```
Comment (zoewangg) [inline test/protocol-tests/src/test/java/software/amazon/awssdk/protocol/tests/connection/SyncClientConnectionInterruptionTest.java:196]: How is this different from `connectionPoolsGetsReusedWhenInterruptedWith_1_MaxConnection`?

File: test/protocol-tests/src/test/java/software/amazon/awssdk/protocol/tests/connection/SyncClientConnectionInterruptionTest.java [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+        toBeInterruptedFuture.cancel(true);
+        // Make sure thread start the Http connections
+        unInterruptedSleep(50);
+        AllTypesResponse allTypesResponse = client.allTypes();
+        assertThat(allTypesResponse.stringMember()).isEqualTo("resultString");
+        executorService.shutdownNow();
+    }
+
+    @Test
+    void connectionPoolsGetsReusedWhenInterruptedWith_Multiple_MaxConnection() throws Exception {
```
Comment (zoewangg) [inline test/protocol-tests/src/test/java/software/amazon/awssdk/protocol/tests/connection/SyncClientConnectionInterruptionTest.java:89]: Do we actually need this test? It seems we have verified the behavior (connection gets released back to the pool) in `connectionPoolsGetsReusedWhenInterruptedWith_1_MaxConnection`

Comment (joviegas) [reply test/protocol-tests/src/test/java/software/amazon/awssdk/protocol/tests/connection/SyncClientConnectionInterruptionTest.java:89]: removing this test case

File: test/protocol-tests/src/test/java/software/amazon/awssdk/protocol/tests/connection/SyncClientConnectionInterruptionTest.java (hunk 4 above)
Comment (joviegas) [reply test/protocol-tests/src/test/java/software/amazon/awssdk/protocol/tests/connection/SyncClientConnectionInterruptionTest.java]: This was added to test default settings with UrlConnection , its no longer needed.

File: test/protocol-tests/src/test/java/software/amazon/awssdk/protocol/tests/connection/SyncClientConnectionInterruptionTest.java [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+    @Test
+    void connectionPoolsGetsReusedWhenInterruptedWith_1_MaxConnection() throws Exception {
+        Integer LONG_DELAY = 1500;
+
+        String urlRegex = "/2016-03-11/allTypes";
+        stubPostRequest(urlRegex, aResponse().withFixedDelay(LONG_DELAY), SAMPLE_BODY);
+        SdkHttpClient httpClient = ApacheHttpClient.builder().maxConnections(1).build();
+        ProtocolRestJsonClient client = getClient(httpClient, Duration.ofMillis(2L * LONG_DELAY)).build();
+
+        ExecutorService executorService = Executors.newFixedThreadPool(5);
```
[the remark this answers is not in the document]
Comment (joviegas) [reply test/protocol-tests/src/test/java/software/amazon/awssdk/protocol/tests/connection/SyncClientConnectionInterruptionTest.java]: done

File: test/protocol-tests/src/test/java/software/amazon/awssdk/protocol/tests/connection/SyncClientConnectionInterruptionTest.java [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+
+    @BeforeEach
+    public void setup() {
+        mockServer.start();
+        stubPostRequest(".*", aResponse(), "{}");
+    }
+
+    @AfterEach
+    public void cleanUp(){
+        executorService.shutdownNow();
```
Comment (zoewangg) [inline test/protocol-tests/src/test/java/software/amazon/awssdk/protocol/tests/connection/SyncClientConnectionInterruptionTest.java:71]: Maybe `@AfterClass`? Tests will probably fail if we shut it down for a single test.

File: test/protocol-tests/src/test/java/software/amazon/awssdk/protocol/tests/connection/SyncClientConnectionInterruptionTest.java [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+/**
+ * Tests to verify Interruption of Threads while Http Connection is in progress to make sure Resources are released.
+ */
+class SyncClientConnectionInterruptionTest {
+    public static final String SAMPLE_BODY = "{\"StringMember"
+                                             + "\":\"resultString\"}";
+    private final WireMockServer mockServer = new WireMockServer(new WireMockConfiguration()
+                                                                     .bindAddress("localhost").dynamicPort());
+
+    ExecutorService executorService = Executors.newCachedThreadPool();
```
Comment (zoewangg) [inline test/protocol-tests/src/test/java/software/amazon/awssdk/protocol/tests/connection/SyncClientConnectionInterruptionTest.java:60]: nit, `private static`

**Review by zoewangg — APPROVED** [review]

File: test/protocol-tests/src/test/java/software/amazon/awssdk/protocol/tests/connection/SyncClientConnectionInterruptionTest.java (hunk 7 above)
Comment (joviegas) [reply test/protocol-tests/src/test/java/software/amazon/awssdk/protocol/tests/connection/SyncClientConnectionInterruptionTest.java:71]: done

**Pull request merged**